File: 2016/2/2b.c
#include <stdio.h>
#include <string.h>
int main(){
	FILE *ptr;
	int c, i=0, j=0;
	char input[10][5000];
	char keys[5][5] = {
		{'x', 'x', '1', 'x', 'x'},
		{'x', '2', '3', '4', 'x'},
		{'5', '6', '7', '8', '9'},
		{'x', 'A', 'B', 'C', 'x'},
		{'x', 'x', 'D', 'x', 'x'}
	};
	ptr = fopen("input", "r");
	while((c = fgetc(ptr)) != EOF){
		if(c != '\n'){
			input[j][i] = c;
			i++;
		}else{
			input[j][i] = '\n';
			i = 0;
			j++;
		}
	}
	int x = 0, y = 2;
	int nLines = j;
	for(int line = 0; line < nLines; line++){
		for(int cmd = 0; cmd < strlen(input[line]); cmd++){
			switch(input[line][cmd]){
				case 'U':
					if(y > 0 && (keys[y-1][x] != 'x')) y--;
					break;
				case 'D':
					if(y < 4 && (keys[y+1][x] != 'x')) y++;
					break;
				case 'L':
					if(x > 0 && (keys[y][x-1] != 'x')) x--;
					break;
				case 'R':
					if(x < 4 && (keys[y][x+1] != 'x')) x++;
					break;
				case '\n':
					printf("%c", keys[y][x]);
					break;
			}
		}
	}
	return 0;
}

File: 2016/2/2b.rb
keys = [['x', 'x', 1, 'x', 'x'], ['x', 2, 3, 4, 'x'], [5, 6, 7, 8, 9], ['x', 'A', 'B', 'C', 'x'], ['x', 'x', 'D', 'x', 'x']]
$x = 0
$y = 2
input = File.readlines("input").map{|l| l.chomp.split("")}
output = Array.new
input.each do |line|
	line.each do |cmd|
		case cmd
		when "U"
			$y -= 1 if($y > 0 && keys[$y - 1][$x] != 'x')
		when "D"
			$y += 1 if($y < 4 && keys[$y + 1][$x] != 'x')
		when "L"
			$x -= 1 if($x > 0 && keys[$y][$x - 1] != 'x')
		when "R"
			$x += 1 if($x < 4 && keys[$y][$x + 1] != 'x')
		end
	end
	output << keys[$y][$x]
end
puts output.join
